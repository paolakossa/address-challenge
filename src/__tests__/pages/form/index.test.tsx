import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { useRouter } from "next/router";
import { FormProvider, useForm } from "react-hook-form";
import FormPage from "@/pages/form";

jest.mock("next/router", () => ({
  useRouter: jest.fn(),
}));

jest.mock("../../../components/CepField", () => ({
  __esModule: true,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  default: ({ name, label, onCepSearch }: any) => {
    return (
      <div>
        {label && <label htmlFor={name}>{label}</label>}
        <input
          id={name}
          data-testid={name}
          onChange={() =>
            onCepSearch?.({
              logradouro: "Rua Teste",
              bairro: "Bairro Teste",
              localidade: "Cidade Teste",
              uf: "TT",
            })
          }
        />
      </div>
    );
  },
}));

const Wrapper = ({ children }: { children: React.ReactNode }) => {
  const methods = useForm();
  return <FormProvider {...methods}>{children}</FormProvider>;
};

describe("FormPage", () => {
  it("should render correctly", () => {
    render(
      <Wrapper>
        <FormPage />
      </Wrapper>
    );

    expect(
      screen.getByText("Digite o CEP para preencher os campos automaticamente.")
    ).toBeInTheDocument();
    expect(screen.getByText("Salvar")).toBeInTheDocument();
  });

  it("must fill in the fields correctly when searching for a zip code", async () => {
    render(
      <Wrapper>
        <FormPage />
      </Wrapper>
    );

    const cepInput = screen.getByTestId("cep");
    fireEvent.change(cepInput, { target: { value: "74070010" } });

    await waitFor(() => {
      expect(screen.getByLabelText("Logradouro:")).toHaveValue("Rua Teste");
      expect(screen.getByLabelText("Bairro:")).toHaveValue("Bairro Teste");
      expect(screen.getByLabelText("Localidade:")).toHaveValue("Cidade Teste");
      expect(screen.getByLabelText("UF:")).toHaveValue("TT");
    });
  });

  it("should store the address in localStorage and redirect on save", async () => {
    const pushMock = jest.fn();
    (useRouter as jest.Mock).mockReturnValue({ push: pushMock });

    render(
      <Wrapper>
        <FormPage />
      </Wrapper>
    );

    fireEvent.change(screen.getByTestId("cep"), {
      target: { value: "74070010" },
    });
    fireEvent.change(screen.getByLabelText("Logradouro:"), {
      target: { value: "Rua Teste" },
    });
    fireEvent.change(screen.getByLabelText("Bairro:"), {
      target: { value: "Bairro Teste" },
    });
    fireEvent.change(screen.getByLabelText("Localidade:"), {
      target: { value: "Cidade Teste" },
    });
    fireEvent.change(screen.getByLabelText("UF:"), { target: { value: "TT" } });

    fireEvent.click(screen.getByText("Salvar"));

    await waitFor(() => {
      expect(localStorage.getItem("addresses")).toContain("Rua Teste");
      expect(pushMock).toHaveBeenCalledWith("/address-list");
    });
  });

  it("must match the snapshot", () => {
    const { asFragment } = render(
      <Wrapper>
        <FormPage />
      </Wrapper>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
