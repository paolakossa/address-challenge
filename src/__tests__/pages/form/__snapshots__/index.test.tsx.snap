// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormPage must match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="h-screen flex flex-col md:flex-row items-center justify-center  bg-[#f4fdf8]"
  >
    <div
      class="w-full md:w-1/2 h-150 md:h-screen bg-cover bg-center md:bg-right"
      style="background-image: url(/assets/images/address.png);"
    />
    <div
      class="md:w-1/2 flex justify-center items-center"
    >
      <div
        class=" bg-gray-50 p-10 rounded-xl shadow-lg max-w-md w-full"
      >
        <h1
          class="text-xl font-bold mb-4 text-center"
        >
           Digite o CEP para preencher os campos automaticamente.
        </h1>
        <form>
          <div>
            <label
              for="cep"
            >
              CEP:
            </label>
            <input
              data-testid="cep"
              id="cep"
            />
          </div>
          <div
            class="mt-2"
          >
            <div>
              <label
                class="block mb-1 font-medium"
                for="logradouro"
              >
                Logradouro:
              </label>
              <div
                class="flex items-center"
              >
                <input
                  class="border p-2 rounded flex-1"
                  id="logradouro"
                  name="logradouro"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="mt-2"
          >
            <div>
              <label
                class="block mb-1 font-medium"
                for="bairro"
              >
                Bairro:
              </label>
              <div
                class="flex items-center"
              >
                <input
                  class="border p-2 rounded flex-1"
                  id="bairro"
                  name="bairro"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="my-2"
          >
            <div>
              <label
                class="block mb-1 font-medium"
                for="localidade"
              >
                Localidade:
              </label>
              <div
                class="flex items-center"
              >
                <input
                  class="border p-2 rounded flex-1"
                  id="localidade"
                  name="localidade"
                  value=""
                />
              </div>
            </div>
          </div>
           
          <div>
            <label
              class="block mb-1 font-medium"
              for="uf"
            >
              UF:
            </label>
            <div
              class="flex items-center"
            >
              <input
                class="border p-2 rounded flex-1"
                id="uf"
                name="uf"
                value=""
              />
            </div>
          </div>
          <button
            class="w-full mt-4 px-4 py-2 bg-[#225741] text-white rounded hover:bg-green-700 transition"
            type="submit"
          >
            Salvar
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
